################################
########## Heading #############
# This tells the terminal to use \emph{bash as the interpreter
#!/bin/bash
# This is a comment. The heading of the script must include
# instructions to the scheduler. The arguments begin with #PBS and
# the options

# First we need to stablish the name of the queue we will utilize.
# According to the queue charts, assigns 8 processors,
# 64GB of RAM and a walltime of one week
#PBS -q batch
# In case you need to specify requirements you can also use this limits
# to set the exact amount of resources to be used. For example 6 processors
# distributed in a single node, 50GB of RAM and 3 days of computing time.
#PBS -l nodes=1:ppn=8
#PBS -l mem=50gb
#PBS -l walltime=72:00:00


# This remaining directives specify the shell to use, our email and notification
# settings as well as the name for our job.
#PBS -S /bin/bash
#PBS -M j.velez248@uniandes.edu.co
#PBS -m abe
#PBS -N bwa

hapmap=/hpcfs/home/ciencias/biologia/cursos/bcom4103/j.velez248/omics/hg19/hapmap_3.3.hg19.vcf
omni=/hpcfs/home/ciencias/biologia/cursos/bcom4103/j.velez248/omics/hg19/1000G_omni2.5.hg19.vcf
g1000snps=/hpcfs/home/ciencias/biologia/cursos/bcom4103/j.velez248/omics/hg19/1000G_phase1.snps.high_confidence.hg19.sites.vcf
dbsnp=/hpcfs/home/ciencias/biologia/cursos/bcom4103/j.velez248/omics/hg19/dbsnp_137.hg19.vcf
mills=/hpcfs/home/ciencias/biologia/cursos/bcom4103/j.velez248/omics/hg19/Mills_and_1000G_gold_standard.indels.hg19.vcf
g1000indels=/hpcfs/home/ciencias/biologia/cursos/bcom4103/j.velez248/omics/hg19/1000G_phase1.indels.hg19.vcf
genoma_ref=/hpcfs/home/ciencias/biologia/cursos/bcom4103/j.velez248/omics/WholeGenomeFasta/genome.fa

#BWA
bwa mem -t8 -M $genoma_ref /hpcfs/home/ciencias/biologia/cursos/bcom4103/j.velez248/FASTQ/34544_S3_L001_R1_001.fastq  /hpcfs/home/ciencias/biologia/cursos/bcom4103/j.velez248/FASTQ/34544_S3_L001_R2_001.fastq > /hpcfs/home/ciencias/biologia/cursos/bcom4103/j.velez248/34544/34544_S3_L001.sam 

#picard1
module load gatk/3.5
module load R

java -jar `which SortSam.jar` INPUT=/hpcfs/home/ciencias/biologia/cursos/bcom4103/j.velez248/34544/34544_S3_L001.sam OUTPUT=/hpcfs/home/ciencias/biologia/cursos/bcom4103/j.velez248/34544/sorted.bam SORT_ORDER=coordinate


module load picardtools/1.140

######################## PICARD AddOrReplaceReadGroups #################

java -jar `which AddOrReplaceReadGroups.jar` \
	I=/hpcfs/home/ciencias/biologia/cursos/bcom4103/j.velez248/34544/sorted.bam/ \
	O=/hpcfs/home/ciencias/biologia/cursos/bcom4103/j.velez248/34544/sorted.rg.bam \
	SM= prueba \
	VALIDATION_STRINGENCY=LENIENT \
	ID=1 \
	LB=unknown \
	PL=illumina \
	PU=1234

######################## PICARD MarkDuplicates #########################
java -Xmx26g -jar `which MarkDuplicates.jar` \
	I=/hpcfs/home/ciencias/biologia/cursos/bcom4103/j.velez248/34544/sorted.rg.bam \
	O=/hpcfs/home/ciencias/biologia/cursos/bcom4103/j.velez248/34544/sorted.rg.md.bam \
	M=/hpcfs/home/ciencias/biologia/cursos/bcom4103/j.velez248/34544/sorted.rg.md.metrics \
	ASSUME_SORTED=true \
	VALIDATION_STRINGENCY=LENIENT

samtools index /hpcfs/home/ciencias/biologia/cursos/bcom4103/j.velez248/34544/sorted.rg.md.bam

##GATK_preprocesamiento

######################## GATK RealignerTargetCreator ###################
java -Xmx12g -jar `which GenomeAnalysisTK.jar` \
    -T RealignerTargetCreator -nt 8 \
    -R $genoma_ref  \
    -I /hpcfs/home/ciencias/biologia/cursos/bcom4103/j.velez248/34544/sorted.rg.md.bam \
    -o /hpcfs/home/ciencias/biologia/cursos/bcom4103/j.velez248/34544/sorted.rg.md.intervals \
    -L /hpcfs/home/ciencias/biologia/cursos/bcom4103/j.velez248/omics/hg19/truseq_exome_targeted_regions.hg19.bed \
    -known /hpcfs/home/ciencias/biologia/cursos/bcom4103/j.velez248/omics/hg19/Mills_and_1000G_gold_standard.indels.hg19.vcf \
    -known /hpcfs/home/ciencias/biologia/cursos/bcom4103/j.velez248/omics/hg19/dbsnp_137.hg19.vcf \
    -known /hpcfs/home/ciencias/biologia/cursos/bcom4103/j.velez248/omics/hg19/1000G_phase1.indels.hg19.vcf
    
    
######################## GATK IndelRealigner ###########################
java -Xmx12g -jar `which GenomeAnalysisTK.jar` \
    -T IndelRealigner \
    -R $genoma_ref \
    -I /hpcfs/home/ciencias/biologia/cursos/bcom4103/j.velez248/34544/sorted.rg.md.bam \
    -targetIntervals /hpcfs/home/ciencias/biologia/cursos/bcom4103/j.velez248/34544/sorted.rg.md.intervals \
    -o /hpcfs/home/ciencias/biologia/cursos/bcom4103/j.velez248/34544/sorted.rg.md.ir.bam \
    --filter_bases_not_stored \
    -known /hpcfs/home/ciencias/biologia/cursos/bcom4103/j.velez248/omics/hg19/Mills_and_1000G_gold_standard.indels.hg19.vcf \
    -known /hpcfs/home/ciencias/biologia/cursos/bcom4103/j.velez248/omics/hg19/dbsnp_137.hg19.vcf \
    -known /hpcfs/home/ciencias/biologia/cursos/bcom4103/j.velez248/omics/hg19/1000G_phase1.indels.hg19.vcf

######################## ANOTHER SAMTOOLS INDEX ########################

samtools index /hpcfs/home/ciencias/biologia/cursos/bcom4103/j.velez248/34544/sorted.rg.md.ir.bam

######################## GATK BaseRecalibrator #########################
java -Xmx12g -jar `which GenomeAnalysisTK.jar` \
    -T BaseRecalibrator -nct 8 \
    -R $genoma_ref \ 
    -I /hpcfs/home/ciencias/biologia/cursos/bcom4103/j.velez248/34544/sorted.rg.md.ir.bam \
    -o /hpcfs/home/ciencias/biologia/cursos/bcom4103/j.velez248/34544/sorted.rg.md.ir.grp \
    -knownSites /hpcfs/home/ciencias/biologia/cursos/bcom4103/j.velez248/omics/hg19/Mills_and_1000G_gold_standard.indels.hg19.vcf \
    -knownSites /hpcfs/home/ciencias/biologia/cursos/bcom4103/j.velez248/omics/hg19/dbsnp_137.hg19.vcf \
    -knownSites /hpcfs/home/ciencias/biologia/cursos/bcom4103/j.velez248/omics/hg19/1000G_phase1.indels.hg19.vcf

######################## GATK PrintReads ###############################
java -Xmx12g -jar `which GenomeAnalysisTK.jar` \
    -T PrintReads -nct 8 \
    -R $genoma_ref \
    -I /hpcfs/home/ciencias/biologia/cursos/bcom4103/j.velez248/34544/sorted.rg.md.ir.bam \
    -BQSR /hpcfs/home/ciencias/biologia/cursos/bcom4103/j.velez248/34544/sorted.rg.md.ir.grp \
    -o /hpcfs/home/ciencias/biologia/cursos/bcom4103/j.velez248/34544/ready.bam


######################## ANOTHER SAMTOOLS INDEX ########################

samtools index /hpcfs/home/ciencias/biologia/cursos/bcom4103/j.velez248/34544/ready.bam

exit 0

# Create a temporary directory locally on the node optimizes I/O performance
TEMP_DIR=/state/partition1/$USER/$PBS_JOBNAME.${PBS_JOBID%.clustermaster.uniandes.edu.co}
# Create the directory to store the computation output. For example
# results inside the working directory.
OUT_DIR=$PBS_O_WORKDIR/results

# Creates a script log with the following information:
SCRIPT_LOG=$PBS_O_WORKDIR/$PBS_JOBNAME.${PBS_JOBID%.clustermaster.uniandes.edu.co}.log
touch $SCRIPT_LOG

# Now store the information into the script log
echo "Job started on" `hostname` `date` >> $SCRIPT_LOG
echo "#####################################" >> $SCRIPT_LOG
echo "PBS working directory: $PBS_O_WORKDIR" >> $SCRIPT_LOG
echo "Current directory:" `pwd` >> $SCRIPT_LOG
echo "Temporary directory: "$TEMP_DIR >> $SCRIPT_LOG
echo "Output directory: "$OUT_DIR >> $SCRIPT_LOG
echo "#####################################" >> $SCRIPT_LOG
echo "User: $PBS_O_LOGNAME" >>$SCRIPT_LOG
echo "Batch job started on $PBS_O_HOST" >>$SCRIPT_LOG
echo "PBS job id: $PBS_JOBID" >>$SCRIPT_LOG
echo "PBS job name: $PBS_JOBNAME" >>$SCRIPT_LOG
echo "PBS environment: $PBS_ENVIRONMENT" >>$SCRIPT_LOG
echo "#####################################" >> $SCRIPT_LOG
echo  >>$SCRIPT_LOG
echo "#####################################" >> $SCRIPT_LOG
echo "#####################################" >> $SCRIPT_LOG
echo "Full Environment:" >> $SCRIPT_LOG
printenv >> $SCRIPT_LOG
echo "#####################################" >> $SCRIPT_LOG
echo  >> $SCRIPT_LOG
echo "The Job is being executed on the following node:" >> $SCRIPT_LOG
cat ${PBS_NODEFILE} >> $SCRIPT_LOG


mkdir -p $TEMP_DIR
mkdir -p $OUT_DIR
cp -Rf $PBS_O_WORKDIR/important_file(s) $TEMP_DIR/.


# We need to load the Gromacs v4.6.5 module
#module load gromacs/4.6.5

# execute gromacs using the configuration file config.txt
cd $TEMP_DIR
grompp config_file


cd $OUT_DIR
mv -f $TEMP_DIR ./
# OPTIONAL (If you selected Body Part II): Output the finishing
# date on the script log
echo "Job Finished: " `date` >> $SCRIPT_LOG
